test/integration/deployment/deployment_test.go

```diff
@@ -134,3 +130,154 @@ func TestDeploymentSelectorImmutability(t *testing.T) {
+	tester.waitForDeploymentRevisionAndImage("1", fakeImage)
+
+	// Make sure the Deployment status becomes valid while manually marking Deployment pods as ready at the same time
+	tester.waitForDeploymentStatusValidAndMarkPodsReady()
+
+	// A new replicaset should be created.
+	tester.expectNewReplicaSet()
+
+	// Pause the deployment.
+	// The paused deployment shouldn't trigger a new rollout.
+	tester.deployment, err = tester.updateDeployment(pauseFn())
+	if err != nil {
+		t.Errorf("failed to pause deployment %s: %v", tester.deployment.Name, err)
+	}
+
+	// Wait for the controller to notice the pause.
+	tester.waitForObservedDeployment(tester.deployment.Generation)
+
+	// Update the deployment template
+	newTGPS := int64(0)
+	tester.deployment, err = tester.updateDeployment(func(update *v1beta1.Deployment) {
+		update.Spec.Template.Spec.TerminationGracePeriodSeconds = &newTGPS
+	})
+	if err != nil {
+		t.Errorf("failed updating deployment %s: %v", tester.deployment.Name, err)
+	}
+
+	// Wait for the controller to notice the rollout.
+	tester.waitForObservedDeployment(tester.deployment.Generation)
+
+	// Verify that the paused deployment won't create new replica set.
+	tester.expectNoNewReplicaSet()
+
+	_, allOldRs, err := deploymentutil.GetOldReplicaSets(tester.deployment, c.ExtensionsV1beta1())
+	if err != nil {
+		t.Errorf("failed retrieving old replicasets of deployment %s: %v", tester.deployment.Name, err)
+	}
+	if len(allOldRs) != 1 {
+		t.Errorf("expected an old replica set, got %v", allOldRs)
+	}
+	if *allOldRs[0].Spec.Template.Spec.TerminationGracePeriodSeconds == newTGPS {
+		t.Errorf("TerminationGracePeriodSeconds on the replica set should be %d, got %d", tgps, newTGPS)
+	}
+}
+
+// Paused deployment can be scaled
+func TestScalePausedDeployment(t *testing.T) {
+	s, closeFn, rm, dc, informers, c := dcSetup(t)
+	defer closeFn()
+	name := "test-scale-paused-deployment"
+	ns := framework.CreateTestingNamespace(name, s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+
+	replicas := int32(1)
+	tester := &deploymentTester{t: t, c: c, deployment: newDeployment(name, ns.Name, replicas)}
+	tgps := int64(1)
+	tester.deployment.Spec.Template.Spec.TerminationGracePeriodSeconds = &tgps
+
+	var err error
+	tester.deployment, err = c.ExtensionsV1beta1().Deployments(ns.Name).Create(tester.deployment)
+	if err != nil {
+		t.Fatalf("failed to create deployment %s: %v", tester.deployment.Name, err)
+	}
+
+	// Start informer and controllers
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	go dc.Run(5, stopCh)
+
+	// Wait for the Deployment to be updated to revision 1
+	tester.waitForDeploymentRevisionAndImage("1", fakeImage)
+
+	// Make sure the Deployment status becomes valid while manually marking Deployment pods as ready at the same time
+	tester.waitForDeploymentStatusValidAndMarkPodsReady()
+
+	// A new replicaset should be created.
+	tester.expectNewReplicaSet()
```

Comment:
So this tester returns errors in some cases, that we always treat as fatal in the call site, and in some other cases like this one, it fatals by itself. Can we make error handling consistent? I would vote for bubbling all errors up to the call site but I don't feel strong either way.

Reply:
At first I chose to fatal in tester because it's much more succinct (1 line vs. 3 lines for each check). However fatal in the call site makes debug easier (it shows you where it failed). For that reason, I'll switch to call site.